## **Import libraries**

In [49]:
import pandas as pd
import numpy as nm
import os

## **Import cvs files**

In [51]:
#path
path = r'/Users/poojathummar/Documents/career Foundary/Data Immersion/Achi 4 Python Fundamentals for Data Analysts/Instacart Basket Analysis'


In [53]:
#orders_products_combined4.6.pkl  from prepared data folder
df_ords_prods_comb= pd.read_pickle(os.path.join(path,'02 Data/Prepared Data/orders_products_combined46.pkl'))

In [55]:
#orders_products_combined4.6.pkl  from prepared data folder
df_prods= pd.read_csv(os.path.join(path,'02 Data/Prepared Data/products_checked45.csv'),index_col = False)

## Step 4:

In [57]:
#to Check the shape of orders_products_combined4.6.pkl
df_ords_prods_comb.shape

(32434489, 12)

In [59]:
df_ords_prods_comb.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id  order_number  \
0             0           0   2539329        1             1   
1             0           0   2539329        1             1   
2             0           0   2539329        1             1   
3             0           0   2539329        1             1   
4             0           0   2539329        1             1   

   orders_day_of_week  order_hour_of_day  days_since_prior_order  product_id  \
0                   2                  8                     NaN         196   
1                   2                  8                     NaN       14084   
2                   2                  8                     NaN       12427   
3                   2                  8                     NaN       26088   
4                   2                  8                     NaN       26405   

   add_to_cart_order  reordered _merge  
0                  1          0   both  
1                  2          0   both  
2                  3          0   both  
3                  4          0   both  
4                  5          0   both

In [61]:
df_ords_prods_comb = df_ords_prods_comb.drop(['Unnamed: 0'], axis=1)

In [63]:
df_ords_prods_comb=df_ords_prods_comb.drop(['Unnamed: 0.1'],axis =1)

In [18]:
df_ords_prods_comb

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2539329        1             1                   2   
2          2539329        1             1                   2   
3          2539329        1             1                   2   
4          2539329        1             1                   2   
...            ...      ...           ...                 ...   
32434484   2977660   206209            13                   1   
32434485   2977660   206209            13                   1   
32434486   2977660   206209            13                   1   
32434487   2977660   206209            13                   1   
32434488   2977660   206209            13                   1   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         8                     NaN       14084   
2                         8                     NaN       12427   
3                         8                     NaN       26088   
4                         8                     NaN       26405   
...                     ...                     ...         ...   
32434484                 12                     7.0       14197   
32434485                 12                     7.0       38730   
32434486                 12                     7.0       31477   
32434487                 12                     7.0        6567   
32434488                 12                     7.0       22920   

          add_to_cart_order  reordered _merge  
0                         1          0   both  
1                         2          0   both  
2                         3          0   both  
3                         4          0   both  
4                         5          0   both  
...                     ...        ...    ...  
32434484                  5          1   both  
32434485                  6          0   both  
32434486                  7          0   both  
32434487                  8          0   both  
32434488                  9          0   both  

[32434489 rows x 10 columns]

In [65]:
#to Check the shape of products_checked.csv
df_prods.shape

(49672, 6)

In [67]:
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [69]:
df_prods = df_prods.drop(['Unnamed: 0'], axis=1)

In [71]:
df_prods

product_id                                       product_name  \
0               1                         Chocolate Sandwich Cookies   
1               2                                   All-Seasons Salt   
2               3               Robust Golden Unsweetened Oolong Tea   
3               4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4               5                          Green Chile Anytime Sauce   
...           ...                                                ...   
49667       49684          Vodka, Triple Distilled, Twist of Vanilla   
49668       49685                 En Croute Roast Hazelnut Cranberry   
49669       49686                                   Artisan Baguette   
49670       49687         Smartblend Healthy Metabolism Dry Cat Food   
49671       49688                             Fresh Foaming Cleanser   

       aisle_id  department_id  prices  
0            61             19     5.8  
1           104             13     9.3  
2            94              7     4.5  
3            38              1    10.5  
4             5             13     4.3  
...         ...            ...     ...  
49667       124              5     5.3  
49668        42              1     3.1  
49669       112              3     7.8  
49670        41              8     4.7  
49671        73             11    13.5  

[49672 rows x 5 columns]

In [73]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

In [75]:
df_prods.dtypes

product_id         int64
product_name      object
aisle_id           int64
department_id      int64
prices           float64
dtype: object

## Step 5:

In [77]:
df_prods['prices'].max()

99999.0

In [79]:
df_ords_prods_comb

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2539329        1             1                   2   
2          2539329        1             1                   2   
3          2539329        1             1                   2   
4          2539329        1             1                   2   
...            ...      ...           ...                 ...   
32434484   2977660   206209            13                   1   
32434485   2977660   206209            13                   1   
32434486   2977660   206209            13                   1   
32434487   2977660   206209            13                   1   
32434488   2977660   206209            13                   1   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         8                     NaN       14084   
2                         8                     NaN       12427   
3                         8                     NaN       26088   
4                         8                     NaN       26405   
...                     ...                     ...         ...   
32434484                 12                     7.0       14197   
32434485                 12                     7.0       38730   
32434486                 12                     7.0       31477   
32434487                 12                     7.0        6567   
32434488                 12                     7.0       22920   

          add_to_cart_order  reordered _merge  
0                         1          0   both  
1                         2          0   both  
2                         3          0   both  
3                         4          0   both  
4                         5          0   both  
...                     ...        ...    ...  
32434484                  5          1   both  
32434485                  6          0   both  
32434486                  7          0   both  
32434487                  8          0   both  
32434488                  9          0   both  

[32434489 rows x 10 columns]

##### here the "product_id" column is common in both table.but '_merge' column is already exist in df_ords_prods_comb data farme so i just gave different name "merge_indicator" in df_ords_prods_merge data frame.we could also drop the "_merge" column from  df_ords_prods_comb data frame .This step will be batter for data accuracy.

In [81]:
df_ords_prods_comb = df_ords_prods_comb.drop(['_merge'], axis=1)

In [83]:
df_ords_prods_comb

order_id  user_id  order_number  orders_day_of_week  \
0          2539329        1             1                   2   
1          2539329        1             1                   2   
2          2539329        1             1                   2   
3          2539329        1             1                   2   
4          2539329        1             1                   2   
...            ...      ...           ...                 ...   
32434484   2977660   206209            13                   1   
32434485   2977660   206209            13                   1   
32434486   2977660   206209            13                   1   
32434487   2977660   206209            13                   1   
32434488   2977660   206209            13                   1   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         8                     NaN       14084   
2                         8                     NaN       12427   
3                         8                     NaN       26088   
4                         8                     NaN       26405   
...                     ...                     ...         ...   
32434484                 12                     7.0       14197   
32434485                 12                     7.0       38730   
32434486                 12                     7.0       31477   
32434487                 12                     7.0        6567   
32434488                 12                     7.0       22920   

          add_to_cart_order  reordered  
0                         1          0  
1                         2          0  
2                         3          0  
3                         4          0  
4                         5          0  
...                     ...        ...  
32434484                  5          1  
32434485                  6          0  
32434486                  7          0  
32434487                  8          0  
32434488                  9          0  

[32434489 rows x 9 columns]

In [36]:
df_ords_prods_merge =df_ords_prods_comb.merge(df_prods, on = 'product_id', indicator = True)

In [85]:
#df_ords_prods_merge2 = df_prods.merge(df_ords_prods_comb, on = 'product_id', indicator = True)
#df_ords_prods_merge3=pd.merge(df_prods,df_ords_prods_comb ,on ='product_id', how ='inner')

In [87]:
df_ords_prods_merge3

product_id                product_name  aisle_id  department_id  \
0                  1  Chocolate Sandwich Cookies        61             19   
1                  1  Chocolate Sandwich Cookies        61             19   
2                  1  Chocolate Sandwich Cookies        61             19   
3                  1  Chocolate Sandwich Cookies        61             19   
4                  1  Chocolate Sandwich Cookies        61             19   
...              ...                         ...       ...            ...   
32404854       49688      Fresh Foaming Cleanser        73             11   
32404855       49688      Fresh Foaming Cleanser        73             11   
32404856       49688      Fresh Foaming Cleanser        73             11   
32404857       49688      Fresh Foaming Cleanser        73             11   
32404858       49688      Fresh Foaming Cleanser        73             11   

          prices  order_id  user_id  order_number  orders_day_of_week  \
0            5.8   3139998      138            28                   6   
1            5.8   1977647      138            30                   6   
2            5.8    389851      709             2                   0   
3            5.8    652770      764             1                   3   
4            5.8   1813452      764             3                   4   
...          ...       ...      ...           ...                 ...   
32404854    13.5   1788356   200215             2                   0   
32404855    13.5   3401313   200377             1                   4   
32404856    13.5    809510   200873             5                   3   
32404857    13.5   2359893   200873             9                   3   
32404858    13.5   2385091   205926            11                   1   

          order_hour_of_day  days_since_prior_order  add_to_cart_order  \
0                        11                     3.0                  5   
1                        17                    20.0                  1   
2                        21                     6.0                 20   
3                        13                     NaN                 10   
4                        17                     9.0                 11   
...                     ...                     ...                ...   
32404854                  9                     5.0                 27   
32404855                 11                     NaN                  5   
32404856                  8                    15.0                 12   
32404857                 15                     5.0                 11   
32404858                 15                     6.0                 35   

          reordered  
0                 0  
1                 1  
2                 0  
3                 0  
4                 1  
...             ...  
32404854          0  
32404855          0  
32404856          0  
32404857          1  
32404858          0  

[32404859 rows x 13 columns]

In [322]:
# Check the output
df_ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     NaN         196                  1          0   
1                     NaN       14084                  2          0   
2                     NaN       12427                  3          0   
3                     NaN       26088                  4          0   
4                     NaN       26405                  5          0   

                              product_name  aisle_id  department_id  prices  \
0                                     Soda        77              7     9.0   
1  Organic Unsweetened Vanilla Almond Milk        91             16    12.5   
2                      Original Beef Jerky        23             19     4.4   
3               Aged White Cheddar Popcorn        23             19     4.7   
4         XL Pick-A-Size Paper Towel Rolls        54             17     1.0   

  _merge  
0   both  
1   both  
2   both  
3   both  
4   both

In [325]:
df_ords_prods_merge['prices'].max()

99999.0

## Step 6:

In [44]:
#Confirm the results of the merge using the merge flag.
df_ords_prods_merge['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

##### This new data frame 'df_ords_pros_comb' has 32,404,859 rows, and each of those rows have information found in both input data sets, as we used an inner join for the purposes of this project.

## Step 7:

In [329]:
# Export data to pkl & csv 
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge46.pkl'))

In [46]:

#df_ords_prods_merge2.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge2-46.pkl'))

In [91]:
#df_ords_prods_merge3.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge3-46.pkl'))